File: tests/support/new-patient.ts
import { expect, type Page } from '@playwright/test';
import { newPatientURL } from './config';
import { clickCenter, enableFlutterAccessibility, textboxNearLabel, typeInto } from './flutter';

export async function openNewPatientForm(page: Page) {
  const newPatientButton = page.getByRole('button', { name: /^(create new patient|new patient)$/i }).first();

  await expect(page.getByText(/Patients List/i).first()).toBeVisible({ timeout: 15000 });
  await expect(newPatientButton).toBeVisible({ timeout: 15000 });

  for (let attempt = 1; attempt <= 3; attempt++) {
    await clickCenter(page, newPatientButton);
    try {
      await page.waitForURL(/\/addPatient(?:$|[?#])/, { timeout: 5000 });
      await enableFlutterAccessibility(page);
      return;
    } catch {
      if (attempt === 3 || /\/addPatient(?:$|[?#])/.test(new URL(page.url()).pathname)) {
        await enableFlutterAccessibility(page);
        return;
      }
    }
  }
}

export async function gotoNewPatientForm(page: Page) {
  await page.goto(newPatientURL);
  await enableFlutterAccessibility(page);
  await expect(page).toHaveURL(/\/newHomeAdmin\?selectedOption=1(?:$|[&#])/);
  await openNewPatientForm(page);
  await expect(page).toHaveURL(/\/addPatient(?:$|[?#])/);
}

export async function expectNewPatientFormVisible(page: Page) {
  await expect(page.getByText(/^Create New Patient$/i).first()).toBeVisible({ timeout: 15000 });
  await expect(page.getByText(/^Patient Information$/i).first()).toBeVisible();
  await expect(page.getByText(/^Name$/i).first()).toBeVisible();
  await expect(page.getByText(/^Patient ID$/i).first()).toBeVisible();
  await expect(page.getByText(/^Gender\*?$/i).first()).toBeVisible();
  await expect(page.getByText(/^Phone Number$/i).first()).toBeVisible();
  await expect(page.getByText(/^Age\*$/i).first()).toBeVisible();
  await expect(page.getByText(/^Address$/i).first()).toBeVisible();
  await expect(page.getByText(/^Registration Date\*$/i).first()).toBeVisible();
  await expect(page.getByRole('button', { name: /^submit$/i }).first()).toBeVisible();
  await expect(page.getByRole('button', { name: /^cancel$/i }).first()).toBeVisible();
}

export type Gender = 'Male' | 'Female' | 'Other';

export type NewPatientRequiredDetails = {
  name: string;
  patientId: string;
  phoneNumber: string;
  address?: string;
  age?: string;
  gender?: Gender;
};

export async function selectGender(page: Page, gender: Gender) {
  await clickCenter(page, page.getByRole('button', { name: /^please select\.\.\.$/i }).first());
  await clickCenter(page, page.getByRole('button', { name: new RegExp(`^${gender}$`, 'i') }).first());
}

export async function fillRequiredNewPatientFields(page: Page, details: NewPatientRequiredDetails) {
  await typeInto(page, await textboxNearLabel(page, /^Name\*?$/i), details.name);
  await selectGender(page, details.gender ?? 'Male');

  // Patient ID defaults to "Auto" mode: typing into it is silently discarded and it resets
  // to the auto-generated value (confirmed live via inputValue() — not a locator bug).
  // Setting a specific id requires switching to "Manual" mode first, which isn't wired up
  // here yet — its radio toggle has no accessible label in this app's semantics tree, so it
  // needs a coordinate-based click. Until that's added, `details.patientId` is accepted but
  // not actually applied; the field keeps its auto-generated value.
  await typeInto(page, await textboxNearLabel(page, /^Patient ID\*?$/i), details.patientId);

  await typeInto(page, await textboxNearLabel(page, /^Phone Number\*?$/i), details.phoneNumber);
  await typeInto(page, await textboxNearLabel(page, /^Address$/i), details.address ?? 'Automation test address');
  await typeInto(page, await textboxNearLabel(page, /^Age\*?$/i), details.age ?? '30');

  await expect(page.getByText(/^Create New Patient$/i).first()).toBeVisible();
}

// Several other flows (inpatient admission, appointment booking, consultations) need a
// patient that's guaranteed not to already be tangled up in some other flow's state from a
// prior run (an existing admission, a pending task, an already-booked slot) -- confirmed live
// as a real, recurring problem: reusing one fixed "test patient" name across many spec files
// eventually collides with whatever state a previous run left that patient in. Creating a
// fresh one per run sidesteps that entirely. The Name field rejects digits ("Invalid text"
// shown live), so uniqueness comes from random letters rather than a timestamp suffix.
function uniquePatientName(label: string): string {
  let suffix = '';
  for (let i = 0; i < 8; i++) suffix += String.fromCharCode(97 + Math.floor(Math.random() * 26));
  return `Automation ${label} ${suffix}`;
}

export async function createPatient(page: Page, label = 'Patient'): Promise<string> {
  const name = uniquePatientName(label);
  await gotoNewPatientForm(page);
  await fillRequiredNewPatientFields(page, {
    name,
    patientId: 'AUTO', // Patient ID defaults to auto-generated mode; this is discarded.
    phoneNumber: String(Date.now()).slice(-10),
  });
  await clickCenter(page, page.getByRole('button', { name: /^submit$/i }).first());
  await page.waitForURL(/\/newHomeAdmin\?selectedOption=1(?:$|[&#])/, { timeout: 15000 });
  return name;
}

File: rust/src/support/new_patient.rs
use crate::support::config;
use crate::support::flutter::{
    click_center, enable_flutter_accessibility, text_matching, text_matching_within, textbox_near_label, type_into,
    wait_for_url_matching,
};
use anyhow::Result;
use playwright_rs::expect;
use playwright_rs::protocol::locator::{AriaRole, GetByRoleOptions};
use playwright_rs::protocol::page::Page;
use regex::Regex;
use std::time::Duration;

pub async fn open_new_patient_form(page: &Page) -> Result<()> {
    let new_patient_button = page.get_by_role(
        AriaRole::Button,
        Some(GetByRoleOptions::default().name("New Patient").exact(false)),
    );

    let patients_list = text_matching_within(page, &Regex::new(r"(?i)Patients List").unwrap(), Duration::from_secs(20)).await?;
    expect(patients_list).with_timeout(Duration::from_secs(15)).to_be_visible().await?;
    expect(new_patient_button.clone()).with_timeout(Duration::from_secs(15)).to_be_visible().await?;

    let add_patient_pattern = Regex::new(r"/addPatient(?:$|[?#])").unwrap();
    for attempt in 1..=3 {
        click_center(page, &new_patient_button).await?;
        let navigated = wait_for_url_matching(page, &add_patient_pattern, Duration::from_secs(5)).await;
        if navigated.is_ok() || add_patient_pattern.is_match(&page.url()) {
            enable_flutter_accessibility(page).await?;
            return Ok(());
        }
        if attempt == 3 {
            enable_flutter_accessibility(page).await?;
            return Ok(());
        }
    }
    Ok(())
}

pub async fn goto_new_patient_form(page: &Page) -> Result<()> {
    page.goto(&config::new_patient_url(), None).await?;
    enable_flutter_accessibility(page).await?;
    wait_for_url_matching(
        page,
        &Regex::new(r"/newHomeAdmin\?selectedOption=1(?:$|[&#])").unwrap(),
        Duration::from_secs(15),
    )
    .await?;
    open_new_patient_form(page).await?;
    wait_for_url_matching(page, &Regex::new(r"/addPatient(?:$|[?#])").unwrap(), Duration::from_secs(15)).await?;
    Ok(())
}

pub async fn expect_new_patient_form_visible(page: &Page) -> Result<()> {
    let checks: &[(&str, u64)] = &[
        (r"(?i)^Create New Patient$", 15),
        (r"(?i)^Patient Information$", 5),
        (r"(?i)^Name$", 5),
        (r"(?i)^Patient ID$", 5),
        (r"(?i)^Gender\*?$", 5),
        (r"(?i)^Phone Number$", 5),
        (r"(?i)^Age\*$", 5),
        (r"(?i)^Address$", 5),
        (r"(?i)^Registration Date\*$", 5),
    ];
    for (pattern, timeout_secs) in checks {
        let locator = text_matching(page, &Regex::new(pattern).unwrap()).await?;
        expect(locator).with_timeout(Duration::from_secs(*timeout_secs)).to_be_visible().await?;
    }

    let submit = page.get_by_role(AriaRole::Button, Some(GetByRoleOptions::default().name("Submit").exact(true)));
    let cancel = page.get_by_role(AriaRole::Button, Some(GetByRoleOptions::default().name("Cancel").exact(true)));
    expect(submit).to_be_visible().await?;
    expect(cancel).to_be_visible().await?;
    Ok(())
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Gender {
    Male,
    Female,
    Other,
}

impl Gender {
    fn as_str(self) -> &'static str {
        match self {
            Gender::Male => "Male",
            Gender::Female => "Female",
            Gender::Other => "Other",
        }
    }
}

pub struct NewPatientRequiredDetails {
    pub name: String,
    pub patient_id: String,
    pub phone_number: String,
    pub address: Option<String>,
    pub age: Option<String>,
    pub gender: Option<Gender>,
}

impl NewPatientRequiredDetails {
    pub fn new(name: impl Into<String>, patient_id: impl Into<String>, phone_number: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            patient_id: patient_id.into(),
            phone_number: phone_number.into(),
            address: None,
            age: None,
            gender: None,
        }
    }
}

pub async fn select_gender(page: &Page, gender: Gender) -> Result<()> {
    let please_select = page
        .get_by_role(
            AriaRole::Button,
            Some(GetByRoleOptions::default().name("Please select...").exact(true)),
        )
        .first();
    click_center(page, &please_select).await?;
    let option = page
        .get_by_role(
            AriaRole::Button,
            Some(GetByRoleOptions::default().name(gender.as_str()).exact(true)),
        )
        .first();
    click_center(page, &option).await?;
    Ok(())
}

pub async fn fill_required_new_patient_fields(page: &Page, details: &NewPatientRequiredDetails) -> Result<()> {
    let name_input = textbox_near_label(page, &Regex::new(r"(?i)^Name\*?$").unwrap()).await?;
    type_into(page, &name_input, &details.name).await?;
    select_gender(page, details.gender.unwrap_or(Gender::Male)).await?;

    // Patient ID defaults to "Auto" mode: typing into it is silently discarded and it resets
    // to the auto-generated value. Setting a specific id requires switching to "Manual" mode
    // first, which isn't wired up here yet -- `details.patient_id` is accepted but not
    // actually applied; the field keeps its auto-generated value.
    let patient_id_input = textbox_near_label(page, &Regex::new(r"(?i)^Patient ID\*?$").unwrap()).await?;
    type_into(page, &patient_id_input, &details.patient_id).await?;

    let phone_input = textbox_near_label(page, &Regex::new(r"(?i)^Phone Number\*?$").unwrap()).await?;
    type_into(page, &phone_input, &details.phone_number).await?;

    let address_input = textbox_near_label(page, &Regex::new(r"(?i)^Address$").unwrap()).await?;
    type_into(page, &address_input, details.address.as_deref().unwrap_or("Automation test address")).await?;

    let age_input = textbox_near_label(page, &Regex::new(r"(?i)^Age\*?$").unwrap()).await?;
    type_into(page, &age_input, details.age.as_deref().unwrap_or("30")).await?;

    let heading = text_matching(page, &Regex::new(r"(?i)^Create New Patient$").unwrap()).await?;
    expect(heading).to_be_visible().await?;
    Ok(())
}

// A patient created ahead of time in the target environment and reused across suites instead
// of creating a fresh one per run.
pub const EXISTING_PATIENT_NAME: &str = "Brad Pitt";

// The Name field rejects digits ("Invalid text" shown live), so uniqueness comes from random
// letters rather than a timestamp suffix.
fn unique_patient_name(label: &str) -> String {
    use rand::RngExt;
    let mut rng = rand::rng();
    let suffix: String = (0..8).map(|_| (b'a' + rng.random_range(0..26)) as char).collect();
    format!("Automation {label} {suffix}")
}

pub async fn create_patient(page: &Page, label: &str) -> Result<String> {
    let name = unique_patient_name(label);
    goto_new_patient_form(page).await?;

    let phone = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)?
        .as_millis()
        .to_string();
    let phone = phone[phone.len().saturating_sub(10)..].to_string();

    let details = NewPatientRequiredDetails::new(name.clone(), "AUTO", phone);
    fill_required_new_patient_fields(page, &details).await?;

    let submit = page.get_by_role(AriaRole::Button, Some(GetByRoleOptions::default().name("Submit").exact(true)));
    click_center(page, &submit).await?;
    wait_for_url_matching(
        page,
        &Regex::new(r"/newHomeAdmin\?selectedOption=1(?:$|[&#])").unwrap(),
        Duration::from_secs(15),
    )
    .await?;
    Ok(name)
}
